subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java

```diff
@@ -0,0 +1,301 @@
+import org.gradle.cache.internal.CacheScopeMapping;
+import org.gradle.cache.internal.InMemoryCacheDecoratorFactory;
+import org.gradle.cache.internal.VersionStrategy;
+import org.gradle.initialization.RootBuildLifecycleListener;
+import org.gradle.initialization.layout.ProjectCacheDir;
+import org.gradle.internal.classloader.ClasspathHasher;
+import org.gradle.internal.classpath.CachedJarFileStore;
+import org.gradle.internal.event.ListenerManager;
+import org.gradle.internal.execution.OutputChangeListener;
+import org.gradle.internal.file.Stat;
+import org.gradle.internal.fingerprint.FileCollectionFingerprinter;
+import org.gradle.internal.fingerprint.FileCollectionFingerprinterRegistry;
+import org.gradle.internal.fingerprint.FileCollectionSnapshotter;
+import org.gradle.internal.fingerprint.GenericFileTreeSnapshotter;
+import org.gradle.internal.fingerprint.classpath.ClasspathFingerprinter;
+import org.gradle.internal.fingerprint.classpath.CompileClasspathFingerprinter;
+import org.gradle.internal.fingerprint.classpath.impl.DefaultClasspathFingerprinter;
+import org.gradle.internal.fingerprint.classpath.impl.DefaultCompileClasspathFingerprinter;
+import org.gradle.internal.fingerprint.impl.AbsolutePathFileCollectionFingerprinter;
+import org.gradle.internal.fingerprint.impl.DefaultFileCollectionFingerprinterRegistry;
+import org.gradle.internal.fingerprint.impl.DefaultFileCollectionSnapshotter;
+import org.gradle.internal.fingerprint.impl.DefaultGenericFileTreeSnapshotter;
+import org.gradle.internal.fingerprint.impl.IgnoredPathFileCollectionFingerprinter;
+import org.gradle.internal.fingerprint.impl.NameOnlyFileCollectionFingerprinter;
+import org.gradle.internal.fingerprint.impl.OutputFileCollectionFingerprinter;
+import org.gradle.internal.fingerprint.impl.RelativePathFileCollectionFingerprinter;
+import org.gradle.internal.hash.DefaultFileHasher;
+import org.gradle.internal.hash.FileHasher;
+import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.hash.StreamHasher;
+import org.gradle.internal.nativeintegration.filesystem.FileSystem;
+import org.gradle.internal.serialize.HashCodeSerializer;
+import org.gradle.internal.service.ServiceRegistration;
+import org.gradle.internal.snapshot.WellKnownFileLocations;
+import org.gradle.internal.vfs.VirtualFileSystem;
+import org.gradle.internal.vfs.impl.DefaultVirtualFileSystem;
+import org.gradle.internal.vfs.impl.RoutingVirtualFileSystem;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.List;
+import java.util.Map;
+
+public class VirtualFileSystemServices extends AbstractPluginServiceRegistry {
+    /**
+     * System property to enable feature preview for partial invalidation.
+     */
+    public static final String VFS_PARTIAL_INVALIDATION_ENABLED_PROPERTY = "org.gradle.unsafe.vfs.partial-invalidation";
+
+    /**
+     * System property to enable feature preview for retaining VFS state between builds.
```

Comment:
```suggestion
     * System property to enable retaining VFS state between builds.
```

Reply:
I don't feel strongly either way. These are feature previews, we just don't use `FeaturePreviews` to enable them. Still, I'm fine with removing the words.

Reply:
I am not sure I know what a feature preview is, at least not any more. It is a feature what we are enabling here, and the feature is experimental.
Therfore, I wouldn't call out `feature preview` in the Javadoc and drop it instead.